File: algorithms/40. Combination Sum II/solution.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()  # use sorted in real case
        comb_arr = []
        comb = []
        def dfs(start, target):
            if target == 0:
                comb_arr.append(tuple(comb))
                return
            for i in range(start, len(candidates)):
                if target < candidates[i]:
                    return
                # we can jump over because c[i] is a subset of c[i-1]
                if i > start and candidates[i] == candidates[i-1]:
                    continue
                comb.append(candidates[i])
                dfs(i+1, target-candidates[i])
                comb.pop()
        dfs(0, target)
        return comb_arr

File: algorithms/40. Combination Sum II/solution.cc
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> path;
        sort(candidates.begin(), candidates.end());
        dfs(0, candidates, target, path, result);
        return result;
    }

    void dfs(int pos, vector<int> & C, int T, vector<int> & path, vector<vector<int>> & res) {
        if ( 0 == T ) {
            res.push_back(path);
            return;
        }
        for ( int i = pos; i < C.size(); ++i ) {
            if ( C[i] > T ) return;
            if ( i > pos && C[i] == C[i-1] ) continue;
            path.push_back(C[i]);
            dfs(i + 1, C, T - C[i], path, res);
            path.pop_back();
        }
    }
};
